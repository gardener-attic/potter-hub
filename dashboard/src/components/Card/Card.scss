.Card,
.Card__inner {
  height: 100%;
  width: 100%;
}

.Card {
  margin: 0 1% 1.5em;
}

.Card a,
.Card a:hover {
  // TODO: See why we need important here - https://github.com/bitnami/hex/issues/217
  // sass-lint:disable no-important
  text-decoration: none !important;
}

.Card__Icon img {
  max-height: 64px;
}

/* max 4 items per row (default) */
@media screen and (min-width: 480px) {
  .Card.Card-responsive {
    width: 48%;
  }
}

@media screen and (min-width: 768px) {
  .Card.Card-responsive {
    width: 31%;
  }
}

@media screen and (min-width: 1100px) {
  .Card.Card-responsive {
    width: 23%;
  }
}

/* max 3 items per row */
@media screen and (min-width: 768px) {
  .Card.Card-responsive-3 {
    width: 48%;
  }
}

@media screen and (min-width: 1100px) {
  .Card.Card-responsive-3 {
    width: 31%;
  }
}

/* max 2 items per row */
@media screen and (min-width: 768px) {
  .Card.Card-responsive-2 {
    width: 48%;
  }
}
