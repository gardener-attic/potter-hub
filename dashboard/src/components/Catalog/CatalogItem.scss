.ListItem__content__info_tag.stable {
  background-color: #1598cb;
}

.ListItem__content__info_tag.incubator {
  background-color: #f58220;
}

.ListItem__content__info_tag_link {
  color: #fff;
  &:hover {
    color: #fff;
  }
  &:visited {
    color: #fff;
  }
}

.ListItem__content__description {
  overflow: hidden;
  height: 80px;
  padding-top: 5px;
  padding-bottom: 5px;
  font-size: 14px;
}
