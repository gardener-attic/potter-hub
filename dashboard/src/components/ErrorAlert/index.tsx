import ErrorAlertHeader from "./ErrorAlertHeader";
import ErrorSelector from "./ErrorSelector";
import MessageAlert from "./MessageAlert";
import NotFoundErrorAlert from "./NotFoundErrorAlert";
import PermissionsErrorAlert from "./PermissionsErrorAlert";
import ServiceBrokersNotFoundAlert from "./ServiceBrokersNotFoundAlert";
import ServiceCatalogNotInstalledAlert from "./ServiceCatalogNotInstalledAlert";
import UnexpectedErrorAlert from "./UnexpectedErrorAlert";

export {
  ErrorAlertHeader,
  ErrorSelector,
  MessageAlert,
  NotFoundErrorAlert,
  PermissionsErrorAlert,
  ServiceBrokersNotFoundAlert,
  ServiceCatalogNotInstalledAlert,
  UnexpectedErrorAlert,
};
