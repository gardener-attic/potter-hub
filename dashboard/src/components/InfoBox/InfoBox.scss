.MessageContainer {
    display: flex;
    align-items: center;
    margin-top: 1em;
}

.error__icon {
    display: inline-block;
    background: #c7c9c8;
    vertical-align: bottom;
    border-radius: 50%;
    height: 30px;
    width: 30px;
}
  
.error__icon > svg {
    margin: auto;
    left: 3px;
    top: 3px;
    color: #ffffff;
    position: relative;
}