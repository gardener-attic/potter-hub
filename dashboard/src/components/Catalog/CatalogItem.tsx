import * as React from "react";

import placeholder from "../../img/placeholder.png";
import { IChart } from "../../shared/types";
import InfoCard from "../InfoCard";

import "./CatalogItem.css";

interface ICatalogItemProps {
  chart: IChart;
  isFeatured?: boolean;
}

// 3 lines description max
const MAX_DESC_LENGTH = 90;

function trimDescription(desc: string): string {
  if (desc.length > MAX_DESC_LENGTH) {
    // Trim to the last word under the max length
    return desc.substr(0, desc.lastIndexOf(" ", MAX_DESC_LENGTH)).concat("...");
  }
  return desc;
}

const CatalogItem: React.SFC<ICatalogItemProps> = props => {
  const { chart, isFeatured } = props;
  const { icon, name, repo } = chart.attributes;
  const iconSrc = icon ? `api/chartsvc/${icon}` : placeholder;
  const latestChartVersion = chart.relationships.latestChartVersion.data.version
  const latestAppVersion = chart.relationships.latestChartVersion.data.app_version;
  const repoTag = <span className="ListItem__content__info_tag_link">{repo.name}</span>;
  const description = (
    <div className="ListItem__content__description">
      {trimDescription(chart.attributes.description)}
    </div>
  );
  const info = <span>
      Chart:&nbsp;{latestChartVersion}<br/>
      App:&nbsp;{latestAppVersion || "-"}
  </span>
  return (
    <InfoCard
      key={`${repo}/${name}`}
      title={name}
      link={`/charts/${chart.id}`}
      info={info}
      icon={iconSrc}
      description={description}
      tag1Content={repoTag}
      tag1Class={repo.name}
      highlight={isFeatured}
      tagContainerCSS={{ display: "flex", alignItems: "center" }}
    />
  );
};

export default CatalogItem;
