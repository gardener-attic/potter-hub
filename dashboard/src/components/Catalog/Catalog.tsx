import { RouterAction } from "connected-react-router";
import * as React from "react";

import { IAppRepositoryState } from "../../reducers/repos";
import { allRepositories } from "../../shared/Catalog";
import { IAppRepository, IChart, IChartState } from "../../shared/types";
import { escapeRegExp, getInfoAnnotation } from "../../shared/utils";
import { CardGrid } from "../Card";
import { ErrorSelector, MessageAlert } from "../ErrorAlert";
import InfoBox from "../InfoBox";
import LoadingWrapper from "../LoadingWrapper";
import PageHeader from "../PageHeader";
import RepoSelector from "../RepoSelector";
import SearchFilter from "../SearchFilter";
import CatalogItem from "./CatalogItem";

interface ICatalogProps {
  appRepoState: IAppRepositoryState;
  charts: IChartState;
  featuredChartIds?: string[];
  filter: string;
  generalRepoInfo?: string;
  repo: string | undefined;
  fetchApprepositories: () => void;
  fetchCharts: (repo: string) => void;
  pushSearchFilter: (filter: string) => RouterAction;
  push: (location: string) => RouterAction;
  defaultRepo: string | undefined;
  targetClusterSecretNamespace: string;
  targetClusterSecretName: string;
  staticCatalogInfo?: string;
}

interface ICatalogState {
  filter: string;
}

class Catalog extends React.Component<ICatalogProps, ICatalogState> {
  public state: ICatalogState = {
    filter: "",
  };

  public componentDidMount() {
    const { filter } = this.props;
    this.setState({
      filter,
    });
    this.props.fetchApprepositories();
    this.fetchCharts();
  }

  public componentDidUpdate(prevProps: ICatalogProps) {
    if (this.props.filter !== prevProps.filter) {
      this.setState({ filter: this.props.filter });
    }

    if (this.props.repo !== prevProps.repo) {
      this.fetchCharts();
    }
  }

  public render() {
    const {
      charts: {
        items: allItems,
        selected: { error: chartError },
      },
      featuredChartIds = [],
      appRepoState: {
        isFetching: areReposFetching,
        errors: { fetch: repoFetchError },
        repos
      },
      charts: { isFetching: areChartsFetching },
      pushSearchFilter,
      repo,
      generalRepoInfo,
    } = this.props;
    if (repoFetchError) {
      return <ErrorSelector error={repoFetchError} action="list" resource="appRepositories" />;
    }

    let selectedApprepo: IAppRepository | undefined
    if (!areReposFetching && repos.length > 0) {
      selectedApprepo = repos.find(r => {
        return r.metadata.name === repo;
      });

      if (repo && !selectedApprepo && repo !== allRepositories.value) {
        return <MessageAlert header={"Cannot find repository. Please check the repository name in the URL."} />;
      }
    }

    let infoMsg = "";
    if (selectedApprepo) {
      infoMsg = getInfoAnnotation(selectedApprepo);
    } else if (this.props.repo === allRepositories.value && generalRepoInfo) {
      infoMsg = generalRepoInfo;
    }

    return (
      <section className="Catalog">
        <LoadingWrapper loaded={!areReposFetching}>
          <PageHeader>{this.renderHeader(pushSearchFilter)}</PageHeader>

          <InfoBox msg={infoMsg} />

          <LoadingWrapper loaded={!areChartsFetching}>
            {this.renderContent(chartError, allItems, featuredChartIds)}
          </LoadingWrapper>
        </LoadingWrapper>
      </section>
    );
  }

  private renderHeader(pushSearchFilter: (filter: string) => RouterAction) {
    const staticCatalogInfo = this.props.staticCatalogInfo ? this.props.staticCatalogInfo : ""
    return (
      <>
        <div style={{ display: "flex", alignItems: "center", width: "100%" }}>
          <h1>Catalog</h1>
          <RepoSelector
            repo={this.props.repo}
            defaultRepo={this.props.defaultRepo}
            push={this.props.push}
            appRepoState={this.props.appRepoState}
          />
          <SearchFilter
            className="margin-l-big"
            placeholder="search charts..."
            onChange={this.handleFilterQueryChange}
            value={this.state.filter}
            onSubmit={pushSearchFilter}
          />
        </div>
        {!this.uiCalledForTargetCluster() && (
          <InfoBox
            cssStyles={{ width: "100%" }}
            className="margin-b-small margin-t-small"
            msg={staticCatalogInfo} />
        )}
      </>
    );
  }

  private renderContent(
    chartError: Error | undefined,
    allItems: IChart[],
    featuredChartIds: string[],
  ) {
    if (chartError) {
      return <MessageAlert header={chartError.message} />;
    }

    const items = this.filteredCharts(allItems, this.state.filter);

    if (items.length === 0) {
      return <MessageAlert header={"No charts found."} />;
    }
    const featuredCharts = [] as React.ReactElement[];
    const unfeaturedCharts = [] as React.ReactElement[];

    items.forEach(item => {
      if (featuredChartIds.includes(item.id)) {
        featuredCharts.push(<CatalogItem isFeatured={true} key={item.id} chart={item} />);
      } else {
        unfeaturedCharts.push(<CatalogItem isFeatured={false} key={item.id} chart={item} />);
      }
    });

    const allCharts = [] as React.ReactElement[];

    allCharts.push(...featuredCharts);
    allCharts.push(...unfeaturedCharts);

    return <CardGrid>{allCharts}</CardGrid>;
  }

  private filteredCharts(charts: IChart[], filter: string) {
    return charts.filter(c => new RegExp(escapeRegExp(filter), "i").test(c.id));
  }

  private handleFilterQueryChange = (filter: string) => {
    this.setState({
      filter,
    });
  };

  private fetchCharts = () => {
    if (this.props.repo) {
      if (this.props.repo === allRepositories.value) {
        this.props.fetchCharts("");
      } else {
        this.props.fetchCharts(this.props.repo);
      }
    }
  }

  private uiCalledForTargetCluster = () => {
    return this.props.targetClusterSecretNamespace !== "" && this.props.targetClusterSecretName !== "";
  }

}

export default Catalog;
