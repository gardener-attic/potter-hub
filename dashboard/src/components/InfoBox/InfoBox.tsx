import * as React from "react";
import { Info } from "react-feather";

import "./InfoBox.css";

export interface IInfoBoxProps {
  msg: string;
  className?: string;
  cssStyles?: React.CSSProperties | undefined;
}

export class InfoBox extends React.Component<IInfoBoxProps, {}> {
  public render() {
    const { msg, cssStyles } = this.props;

    if (msg !== "") {
      return (
        <div className={`alert alert-warning MessageContainer ${this.props.className}`} role="alert" style={cssStyles}>
          <span className="error__icon margin-r-small">
            <Info />
          </span>
          <div dangerouslySetInnerHTML={{ __html: msg }} />
        </div>
      );
    } else {
      return null;
    }
  }
}

export default InfoBox;
