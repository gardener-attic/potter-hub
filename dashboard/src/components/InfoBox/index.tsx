import InfoBox from "./InfoBox";

export default InfoBox;
