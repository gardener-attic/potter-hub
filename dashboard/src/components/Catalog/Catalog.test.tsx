import InfoBox from "components/InfoBox";
import { mount, shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import { IAppRepositoryState } from "../../reducers/repos";
import { IAppRepository, IChart, IChartState } from "../../shared/types";
import { CardGrid } from "../Card";
import { ErrorSelector, MessageAlert } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";
import PageHeader from "../PageHeader";
import SearchFilter from "../SearchFilter";
import Catalog from "./Catalog";
import CatalogItem from "./CatalogItem";

const defaultChartState = {
  isFetching: false,
  selected: {} as IChartState["selected"],
  deployed: {} as IChartState["deployed"],
  items: [],
  updatesInfo: {},
} as IChartState;
const defaultAppRepoState = {
  isFetching: false,
  repos: [
    {
      metadata: {
        name: "stable",
      },
    } as IAppRepository,
  ] as IAppRepository[],
  errors: {},
} as IAppRepositoryState;
const defaultProps = {
  appRepoState: defaultAppRepoState,
  charts: defaultChartState,
  filter: "",
  fetchCharts: jest.fn(),
  pushSearchFilter: jest.fn(),
  repo: "",
  fetchApprepositories: jest.fn(),
  onToggleFeaturedCharts: jest.fn(),
  showFeaturedCharts: true,
  isFeatured: false,
  push: jest.fn(),
  defaultRepo: undefined,
  targetClusterSecretNamespace: "ns-1",
  targetClusterSecretName: "my-cluster.kubeconfig",
  staticCatalogInfo: "info msg when catalog is opened without target cluster"
};

it("propagates the filter from the props", () => {
  const wrapper = shallow(<Catalog {...defaultProps} filter="foo" />);
  expect(wrapper.state("filter")).toBe("foo");
});

it("reloads charts when the repo changes", () => {
  const spy = jest.fn();
  const wrapper = shallow(<Catalog {...defaultProps} fetchCharts={spy} />);
  wrapper.setProps({ repo: "bitnami" });
  expect(spy).toHaveBeenCalledTimes(1);
  expect(spy).toHaveBeenCalledWith("bitnami");
});

it("updates the filter from props", () => {
  const wrapper = shallow(<Catalog {...defaultProps} />);
  wrapper.setProps({ filter: "foo" });
  expect(wrapper.state("filter")).toBe("foo");
});

it("keeps the filter from the state", () => {
  const wrapper = shallow(<Catalog {...defaultProps} />);
  expect(wrapper.state("filter")).toBe("");
  wrapper.setState({ filter: "foo" });
  expect(wrapper.state("filter")).toBe("foo");
});

it("renders a LoadingWrapper when fetching charts", () => {
  const wrapper = shallow(
    <Catalog {...defaultProps} charts={{ ...defaultChartState, isFetching: true }} />,
  );
  const loadingWrapper = wrapper.find(LoadingWrapper);
  expect(loadingWrapper).toExist();
  expect(loadingWrapper.at(0).props().loaded).toEqual(true);
  expect(loadingWrapper.at(1).props().loaded).toEqual(false);
});

it("renders an error when fetching apprepos failed", () => {
  const errMessage = "test123";
  const wrapper = mount(
    <Catalog
      {...defaultProps}
      appRepoState={{
        ...defaultAppRepoState,
        errors: { fetch: { name: "", message: errMessage } },
      }}
    />,
  );
  expect(wrapper.find(ErrorSelector)).toExist();
  expect(wrapper.find(ErrorSelector).text()).toContain(errMessage);
});

it("renders a LoadingWrapper when fetching apprepos", () => {
  const wrapper = mount(
    <Catalog {...defaultProps} appRepoState={{ ...defaultAppRepoState, isFetching: true }} />,
  );
  expect(wrapper.find(LoadingWrapper)).toExist();
  expect(wrapper.find(LoadingWrapper).props().loaded).toEqual(false);
});

describe("renderization", () => {
  context("when no charts", () => {
    it("should render a distinct 'no charts' message", () => {
      const wrapper = shallow(<Catalog {...defaultProps} />);
      expect(wrapper.find(MessageAlert)).toExist();
      // expect(wrapper.find(".Catalog")).not.toExist();
      expect(wrapper.find(MessageAlert).props().header).toEqual("No charts found.");
      expect(wrapper).toMatchSnapshot();
    });
  });

  context("when charts available", () => {
    const chartState = {
      isFetching: false,
      selected: {} as IChartState["selected"],
      items: [
        { id: "foo", attributes: { description: "" } } as IChart,
        { id: "bar", attributes: { description: "" } } as IChart,
      ],
    } as IChartState;

    it("should render the list of charts", () => {
      const wrapper = shallow(<Catalog {...defaultProps} charts={chartState} />);

      expect(wrapper.find(MessageAlert)).not.toExist();
      expect(wrapper.find(PageHeader)).toExist();
      expect(wrapper.find(SearchFilter)).toExist();

      const cardGrid = wrapper.find(CardGrid);
      expect(cardGrid).toExist();
      expect(cardGrid.children().length).toBe(chartState.items.length);
      expect(
        cardGrid
          .children()
          .at(0)
          .props().chart,
      ).toEqual(chartState.items[0]);
      expect(
        cardGrid
          .children()
          .at(1)
          .props().chart,
      ).toEqual(chartState.items[1]);
      expect(wrapper).toMatchSnapshot();
    });

    it("should filter apps", () => {
      // Filter "foo" app
      const wrapper = shallow(<Catalog {...defaultProps} charts={chartState} filter="foo" />);

      const cardGrid = wrapper.find(CardGrid);
      expect(cardGrid).toExist();
      expect(cardGrid.children().length).toBe(1);
      expect(
        cardGrid
          .children()
          .at(0)
          .props().chart,
      ).toEqual(chartState.items[0]);
    });
  });
});

describe("featured charts", () => {
  const charts: IChartState = {
    isFetching: false,
    selected: {} as IChartState["selected"],
    items: [
      { id: "foo", attributes: { description: "" } } as IChart,
      { id: "bar", attributes: { description: "" } } as IChart,
      { id: "flupp", attributes: { description: "" } } as IChart,
    ],
    deployed: {} as IChartState["deployed"],
  };

  it("shows featured catalogItems", () => {
    const featuredChartIds: string[] = ["foo", "bar"];
    const wrapper = shallow(
      <Catalog {...defaultProps} charts={charts} featuredChartIds={featuredChartIds} />,
    );
    expect(
      wrapper
        .find(CatalogItem)
        .at(0)
        .props().isFeatured,
    ).toBeTruthy();
    expect(
      wrapper
        .find(CatalogItem)
        .at(1)
        .props().isFeatured,
    ).toBeTruthy();
    expect(
      !wrapper
        .find(CatalogItem)
        .at(2)
        .props().isFeatured,
    ).toBeTruthy();
  });

  it("not render featured charts if no data", () => {
    const featuredChartIds: string[] = [];
    const wrapper = shallow(
      <Catalog {...defaultProps} charts={charts} featuredChartIds={featuredChartIds} />,
    );
    expect(
      !wrapper
        .find(CatalogItem)
        .at(0)
        .props().isFeatured,
    ).toBeTruthy();
    expect(
      !wrapper
        .find(CatalogItem)
        .at(1)
        .props().isFeatured,
    ).toBeTruthy();
    expect(
      !wrapper
        .find(CatalogItem)
        .at(2)
        .props().isFeatured,
    ).toBeTruthy();
  });
});

describe("staticCatalogInfo message", () => {
  it("doesn't show info message when ui opened for target cluster", () => {
    const wrapper = shallow(<Catalog {...defaultProps} />);
    expect(wrapper.find(InfoBox).length).toEqual(1);
    expect(wrapper.find(InfoBox).at(0).prop("msg")).toEqual("");
  });

  it("shows info message when ui opened without target cluster", () => {
    const wrapper = shallow(<Catalog {...defaultProps} targetClusterSecretName="" targetClusterSecretNamespace="" />);
    expect(wrapper.find(InfoBox).length).toEqual(2);
    expect(wrapper.find(InfoBox).at(0).prop("msg")).toEqual("info msg when catalog is opened without target cluster");
  });
})

