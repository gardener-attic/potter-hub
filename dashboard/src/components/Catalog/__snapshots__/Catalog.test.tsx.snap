// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderization when charts available should render the list of charts 1`] = `
<section
  className="Catalog"
>
  <LoadingWrapper
    loaded={true}
    type={0}
  >
    <PageHeader>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "width": "100%",
          }
        }
      >
        <h1>
          Catalog
        </h1>
        <RepoSelector
          appRepoState={
            Object {
              "errors": Object {},
              "isFetching": false,
              "repos": Array [
                Object {
                  "metadata": Object {
                    "name": "stable",
                  },
                },
              ],
            }
          }
          push={[MockFunction]}
          repo=""
        />
        <SearchFilter
          className="margin-l-big"
          onChange={[Function]}
          onSubmit={[MockFunction]}
          placeholder="search charts..."
          value=""
        />
      </div>
    </PageHeader>
    <InfoBox
      msg=""
    />
    <LoadingWrapper
      loaded={true}
      type={0}
    >
      <CardGrid>
        <CatalogItem
          chart={
            Object {
              "attributes": Object {
                "description": "",
              },
              "id": "foo",
            }
          }
          isFeatured={false}
          key="foo"
        />
        <CatalogItem
          chart={
            Object {
              "attributes": Object {
                "description": "",
              },
              "id": "bar",
            }
          }
          isFeatured={false}
          key="bar"
        />
      </CardGrid>
    </LoadingWrapper>
  </LoadingWrapper>
</section>
`;

exports[`renderization when no charts should render a distinct 'no charts' message 1`] = `
<section
  className="Catalog"
>
  <LoadingWrapper
    loaded={true}
    type={0}
  >
    <PageHeader>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "width": "100%",
          }
        }
      >
        <h1>
          Catalog
        </h1>
        <RepoSelector
          appRepoState={
            Object {
              "errors": Object {},
              "isFetching": false,
              "repos": Array [
                Object {
                  "metadata": Object {
                    "name": "stable",
                  },
                },
              ],
            }
          }
          push={[MockFunction]}
          repo=""
        />
        <SearchFilter
          className="margin-l-big"
          onChange={[Function]}
          onSubmit={[MockFunction]}
          placeholder="search charts..."
          value=""
        />
      </div>
    </PageHeader>
    <InfoBox
      msg=""
    />
    <LoadingWrapper
      loaded={true}
      type={0}
    >
      <MessageAlertPage
        header="No charts found."
      />
    </LoadingWrapper>
  </LoadingWrapper>
</section>
`;
